collections/binary_heap/src/min_heap.rs

```diff
@@ -0,0 +1,197 @@
+//! # MinHeap
+//!
+//! ## Fields
+//! - data : 실제 heap tree가 구성되는 벡터, memory management는 Vec의 method에 일임한다
+//! - comparator : 비교 방법을 정의한 객체
+//!
+//! ## Generic type
+//! ### T
+//! 실제 저장하게 될 데이터의 타입
+//! ### C
+//! Comparator trait을 구현한 임의 타입
+//!
+//! ## Sub Modules
+//!
+//! ### comparator
+//! Comparator trait의 구현체.
+//!
+
+use std::cmp::Ordering;
+use std::ops::{Deref, DerefMut};
+
+use crate::heap_logic::{build_heap, min_heapify, get_parent};
+use crate::comparator::{Comparator, DefaultComparator};
+
+pub struct MinHeap<T, C: Comparator<T>> {
+	data: Vec<T>,
+	comparator: C,
+}
+
+impl<T, C> MinHeap<T, C>
+where
+	C : Comparator<T>
+{
+	/// create empty min heap
+	pub fn new(comp : C) -> MinHeap<T, C> {
+		MinHeap {
+			data : Vec::new(),
+			comparator : comp,
+		}
+	}
+
+	/// create min heap with vector
+	pub fn from_vec(mut vec : Vec<T>, comp : C) -> MinHeap<T, C> {
+		let comparator = comp;
+		build_heap(&mut vec, &comparator);
+		MinHeap {
+			data : vec,
+			comparator,
+		}
+	}
+
+	/// push new element to the heap
+	pub fn push(&mut self, elem : T) {
+		let data = &mut self.data;
+		data.push(elem);
+		let mut cur_idx = data.len() - 1;
+		let mut parent_idx = get_parent(cur_idx);
+		while parent_idx < data.len()
+			&& cur_idx < data.len()
+			&& Ordering::Greater == self.comparator.compare(&data[parent_idx], &data[cur_idx])
+		{
+			data.swap(parent_idx, cur_idx);	// pull up
+			cur_idx = parent_idx;
+			parent_idx = get_parent(parent_idx);
+		}
+	}
+
+	/// add multiple element to the heap
+	pub fn extend(&mut self, elems : &mut Vec<T>) {
+		let data = &mut self.data;
+		data.append(elems);
+		build_heap(data,&self.comparator);	// O(n)
+	}
+
+	/// extract ownership of the element at root
+	/// O(log n)
+	pub fn pop(&mut self) -> Option<T> {
+		if self.is_empty() {
+			return None;
+		}
+		let data = &mut self.data;
+		let end_idx = data.len() - 1;	// already checked the emptiness, no overflow
+		data.swap(0, end_idx);
+		let result = data.pop();
+		min_heapify(data, &self.comparator, 0);	// O(log n)
+		result
+	}
+
+	/// # Description
+	/// get mutable reference of root of binary heap
+	/// it's source will be heaped when the PeekMut drops
+	pub fn peek_mut<'a>(&'a mut self) -> Option<PeekMut<'a, T, C>> {
+		if !self.is_empty() {
+			return Some(PeekMut {
+				comp : &self.comparator,
+				source : &mut self.data,
+			});
+		}
+		None
+	}
+
+	pub fn len(&self) -> usize {
+		self.data.len()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.data.is_empty()
+	}
+
+	pub fn top(&self) -> Option<&T> {
+		if self.data.is_empty() {
+			return None;
+		}
+		Some(&self.data[0])
+	}
```

Comment:
```suggestion
	pub fn top(&self) -> Option<&T> {
            self.data.get(0)
	}
```
slice::get메서드를 활용하면 스타일이 조금더 나은 것 같습니다.

Reply:
확실히 코드가 간결한게, 수정했습니다.